File: apps/api/app/rag/reranker.py
from __future__ import annotations

import re

from .types import GroundingPassage

_TERM = re.compile(r"\w+", re.UNICODE)


def _terms(value: str) -> set[str]:
    return set(_TERM.findall(value.casefold()))


def rerank_candidates(
    query: str,
    candidates: list[GroundingPassage],
    max_candidates: int,
) -> list[GroundingPassage]:
    bounded = candidates[:max_candidates]
    query_terms = _terms(query)
    ranked = []
    for index, passage in enumerate(bounded):
        overlap = len(query_terms & _terms(passage.text))
        ranked.append((passage, index, overlap))
    ranked.sort(key=lambda item: (-item[2], -item[0].score, item[1]))
    return [passage for passage, _, _ in ranked]

File: apps/web/lib/rag/reranker.ts
import type { GroundingPassage } from "@/lib/rag/types";

export type Reranker = (
  query: string,
  candidates: GroundingPassage[],
  maxCandidates: number,
  signal?: AbortSignal,
) => Promise<GroundingPassage[]>;

function terms(value: string): Set<string> {
  return new Set(value.toLocaleLowerCase().match(/[\p{L}\p{N}]+/gu) ?? []);
}

export async function rerankCandidates(
  query: string,
  candidates: GroundingPassage[],
  maxCandidates: number,
  signal?: AbortSignal,
): Promise<GroundingPassage[]> {
  signal?.throwIfAborted();
  const bounded = candidates.slice(0, maxCandidates);
  const queryTerms = terms(query);

  return bounded
    .map((passage, index) => {
      const passageTerms = terms(passage.text);
      const overlap = [...queryTerms].filter((term) => passageTerms.has(term)).length;
      return { passage, index, overlap };
    })
    .sort(
      (left, right) =>
        right.overlap - left.overlap ||
        right.passage.score - left.passage.score ||
        left.index - right.index,
    )
    .map(({ passage }) => passage);
}
